File: 5.问题五贪心+遗传/optimizer.cpp
#include "optimizer.hpp"
#include <iostream>
#include <algorithm>
#include <limits>
#include <cmath>
#include <set>
#include <omp.h>

namespace Optimizer {

// ObscurationOptimizer Implementation
ObscurationOptimizer::ObscurationOptimizer(
    const std::string& missile_id, 
    const std::unordered_map<std::string, int>& uav_assignments)
    : uav_assignments_(uav_assignments)
    , time_step_(0.1)
    , rng_(std::random_device{}())
{
    missile_ = std::make_unique<CoreObjects::Missile>(missile_id);
    target_ = std::make_unique<CoreObjects::TargetCylinder>(Config::TRUE_TARGET_SPECS);
    target_key_points_ = target_->get_key_points();
}

std::pair<StrategyMap, double> ObscurationOptimizer::solve(
    const std::vector<Bounds>& bounds, 
    const DESettings& settings)
{
    auto [optimal_vars, max_time] = differential_evolution(bounds, settings);
    
    // 重新构建最优策略用于详细输出
    StrategyMap optimal_strategy = parse_decision_variables(optimal_vars);
    return {optimal_strategy, -max_time}; // 注意取负号，因为我们最小化负值
}

double ObscurationOptimizer::objective_function(const VectorXd& decision_variables) {
    try {
        StrategyMap strategies = parse_decision_variables(decision_variables);
        
        std::vector<std::unique_ptr<CoreObjects::SmokeCloud>> smoke_clouds;
        
        for (const auto& [uav_id, uav_strat] : strategies) {
            auto uav = std::make_unique<CoreObjects::UAV>(uav_id);
            uav->set_flight_strategy(uav_strat.speed, uav_strat.angle);
            
            for (const auto& g_strat : uav_strat.grenades) {
                auto grenade = uav->deploy_grenade(g_strat.t_deploy, g_strat.t_fuse);
                smoke_clouds.push_back(grenade->generate_smoke_cloud());
            }
        }
        
        if (smoke_clouds.empty()) {
            return 0.0;
        }
        
        // 计算仿真时间范围
        double sim_start_time = std::numeric_limits<double>::max();
        double sim_end_time = std::numeric_limits<double>::lowest();
        
        for (const auto& cloud : smoke_clouds) {
            sim_start_time = std::min(sim_start_time, cloud->get_start_time());
            sim_end_time = std::max(sim_end_time, cloud->get_end_time());
        }
        
        // 使用集合去重计算有效遮蔽时间点
        std::set<int> obscured_time_points;
        
        for (double t = sim_start_time; t < sim_end_time; t += time_step_) {
            // 获取当前所有有效的云团中心
            std::vector<Vector3d> active_cloud_centers;
            
            for (const auto& cloud : smoke_clouds) {
                auto center = cloud->get_center(t);
                if (center.has_value()) {
                    active_cloud_centers.push_back(center.value());
                }
            }
            
            if (active_cloud_centers.empty()) {
                continue;
            }
            
            // 调用协同判断函数
            Vector3d missile_pos = missile_->get_position(t);
            if (Geometry::check_collective_obscuration(
                    missile_pos, 
                    active_cloud_centers, 
                    target_key_points_)) {
                obscured_time_points.insert(static_cast<int>(std::round(t / time_step_)));
            }
        }
        
        double total_time = obscured_time_points.size() * time_step_;
        return -total_time; // 返回负值用于最小化
        
    } catch (const std::exception&) {
        return 0.0; // 无效策略，返回最差分数
    }
}

std::pair<VectorXd, double> ObscurationOptimizer::differential_evolution(
    const std::vector<Bounds>& bounds,
    const DESettings& settings)
{
    return DifferentialEvolution::optimize(
        [this](const VectorXd& x) { return this->objective_function(x); },
        bounds,
        settings
    );
}

// DifferentialEvolution Implementation
std::pair<VectorXd, double> DifferentialEvolution::optimize(
    ObjectiveFunction objective,
    const std::vector<Bounds>& bounds,
    const DESettings& settings)
{
    const int dim = bounds.size();
    std::mt19937 rng(std::random_device{}());
    
    // 初始化种群
    auto population = initialize_population(bounds, settings.population_size, rng);
    std::vector<double> fitness(settings.population_size);
    
    // 设置OpenMP线程数
    int num_threads = settings.num_threads;
    if (num_threads == -1) {
        num_threads = omp_get_max_threads();
    }
    omp_set_num_threads(num_threads);
    
    // 评估初始种群
    #pragma omp parallel for
    for (int i = 0; i < settings.population_size; ++i) {
        fitness[i] = objective(population[i]);
    }
    
    // 找到最佳个体
    auto best_it = std::min_element(fitness.begin(), fitness.end());
    int best_idx = std::distance(fitness.begin(), best_it);
    VectorXd best_individual = population[best_idx];
    double best_fitness = *best_it;
    
    if (settings.verbose) {
        std::cout << "DE初始化完成，种群大小: " << settings.population_size 
                  << ", 线程数: " << num_threads 
                  << ", 初始最佳适应度: " << -best_fitness << std::endl;
    }
    
    // 主进化循环
    for (int iteration = 0; iteration < settings.max_iterations; ++iteration) {
        std::vector<VectorXd> trial_population(settings.population_size);
        std::vector<double> trial_fitness(settings.population_size);
        
        // 生成试验向量
        #pragma omp parallel for
        for (int i = 0; i < settings.population_size; ++i) {
            std::mt19937 local_rng(rng() + i); // 每个线程独立的随机数生成器
            trial_population[i] = mutate_and_crossover(
                population, i, bounds, 
                settings.differential_weight, 
                settings.crossover_rate, 
                local_rng
            );
            trial_fitness[i] = objective(trial_population[i]);
        }
        
        // 选择操作
        bool improved = false;
        for (int i = 0; i < settings.population_size; ++i) {
            if (trial_fitness[i] < fitness[i]) {
                population[i] = trial_population[i];
                fitness[i] = trial_fitness[i];
                
                if (trial_fitness[i] < best_fitness) {
                    best_individual = trial_population[i];
                    best_fitness = trial_fitness[i];
                    improved = true;
                }
            }
        }
        
        // 输出进度
        if (settings.verbose && (iteration % 50 == 0 || improved)) {
            std::cout << "迭代 " << iteration << ", 最佳适应度: " << -best_fitness << std::endl;
        }
        
        // 收敛检查
        if (improved && std::abs(best_fitness) < settings.tolerance) {
            if (settings.verbose) {
                std::cout << "收敛于迭代 " << iteration << std::endl;
            }
            break;
        }
    }
    
    if (settings.verbose) {
        std::cout << "优化完成，最终适应度: " << -best_fitness << std::endl;
    }
    
    return {best_individual, best_fitness};
}

VectorXd DifferentialEvolution::generate_random_individual(
    const std::vector<Bounds>& bounds,
    std::mt19937& rng)
{
    VectorXd individual(bounds.size());
    
    for (size_t i = 0; i < bounds.size(); ++i) {
        std::uniform_real_distribution<double> dist(bounds[i].lower, bounds[i].upper);
        individual[i] = dist(rng);
    }
    
    return individual;
}

std::vector<VectorXd> DifferentialEvolution::initialize_population(
    const std::vector<Bounds>& bounds,
    int population_size,
    std::mt19937& rng)
{
    std::vector<VectorXd> population;
    population.reserve(population_size);
    
    for (int i = 0; i < population_size; ++i) {
        population.push_back(generate_random_individual(bounds, rng));
    }
    
    return population;
}

VectorXd DifferentialEvolution::mutate_and_crossover(
    const std::vector<VectorXd>& population,
    int target_idx,
    const std::vector<Bounds>& bounds,
    double differential_weight,
    double crossover_rate,
    std::mt19937& rng)
{
    const int population_size = population.size();
    const int dim = bounds.size();
    
    // 随机选择三个不同的个体 (排除目标个体)
    std::vector<int> candidates;
    for (int i = 0; i < population_size; ++i) {
        if (i != target_idx) {
            candidates.push_back(i);
        }
    }
    
    std::shuffle(candidates.begin(), candidates.end(), rng);
    
    int a = candidates[0];
    int b = candidates[1];
    int c = candidates[2];
    
    // 变异: V = X_a + F * (X_b - X_c)
    VectorXd mutant = population[a] + differential_weight * (population[b] - population[c]);
    ensure_bounds(mutant, bounds);
    
    // 交叉
    VectorXd trial = population[target_idx];
    std::uniform_real_distribution<double> uniform(0.0, 1.0);
    std::uniform_int_distribution<int> rand_dim(0, dim - 1);
    
    int random_dim = rand_dim(rng); // 确保至少有一个维度被交叉
    
    for (int i = 0; i < dim; ++i) {
        if (uniform(rng) < crossover_rate || i == random_dim) {
            trial[i] = mutant[i];
        }
    }
    
    ensure_bounds(trial, bounds);
    return trial;
}

void DifferentialEvolution::ensure_bounds(VectorXd& individual, const std::vector<Bounds>& bounds) {
    for (size_t i = 0; i < bounds.size(); ++i) {
        individual[i] = std::clamp(individual[i], bounds[i].lower, bounds[i].upper);
    }
}

// GlobalOptimizer Implementation
GlobalOptimizer::GlobalOptimizer(const std::vector<std::string>& uav_ids,
                                 const std::vector<std::string>& missile_ids,
                                 const std::unordered_map<std::string, double>& threat_weights,
                                 const std::unordered_map<std::string, int>& uav_grenade_counts)
    : uav_ids_(uav_ids)
    , missile_ids_(missile_ids)
    , threat_weights_(threat_weights)
    , uav_grenade_counts_(uav_grenade_counts)
    , target_(Config::TRUE_TARGET_SPECS)
    , time_step_(0.1)
    , num_missiles_(missile_ids.size())
{
    for (const auto& id : uav_ids_) {
        uavs_.emplace(id, CoreObjects::UAV(id));
    }
    for (const auto& id : missile_ids_) {
        missiles_.emplace(id, CoreObjects::Missile(id));
    }
    target_key_points_ = target_.get_key_points();
}

std::pair<StrategyMap, double> GlobalOptimizer::solve(const std::vector<Bounds>& bounds, 
                                                      const DESettings& settings) {
    
    ObjectiveFunction obj_func = [this](const VectorXd& dv) -> double {
        return this->objective_function_impl(dv);
    };
    
    auto [optimal_vars, min_score] = DifferentialEvolution::optimize(obj_func, bounds, settings);
    double max_score = -min_score;
    
    StrategyMap optimal_strategy = parse_decision_variables(optimal_vars);
    
    return {optimal_strategy, max_score};
}

StrategyMap GlobalOptimizer::parse_decision_variables(const VectorXd& decision_variables) {
    StrategyMap strategy;
    int dv_index = 0;
    
    for (const auto& uav_id : uav_ids_) {
        int num_grenades = uav_grenade_counts_.at(uav_id);
        UAVStrategy uav_strat;
        
        uav_strat.speed = decision_variables[dv_index++];
        uav_strat.angle = decision_variables[dv_index++];
        
        double last_td = 0.0;
        for (int i = 0; i < num_grenades; ++i) {
            UAVStrategy::GrenadeDeployment g_strat;
            
            double t_d_or_delta = decision_variables[dv_index++];
            g_strat.t_fuse = decision_variables[dv_index++];
            double target_selector = decision_variables[dv_index++];
            
            g_strat.t_deploy = (i == 0) ? t_d_or_delta : last_td + t_d_or_delta;
            last_td = g_strat.t_deploy;
            
            int target_missile_index = std::min(static_cast<int>(target_selector * num_missiles_), num_missiles_ - 1);
            g_strat.target_missile = missile_ids_.at(target_missile_index);
            
            uav_strat.grenades.push_back(g_strat);
        }
        strategy[uav_id] = uav_strat;
    }
    return strategy;
}

double GlobalOptimizer::objective_function_impl(const VectorXd& decision_variables) {
    StrategyMap strategy;
    try {
        strategy = parse_decision_variables(decision_variables);
    } catch (const std::exception& e) {
        return 1e9;
    }

    std::vector<std::unique_ptr<CoreObjects::SmokeCloud>> all_smoke_clouds;
    
    // 按无人机生成所有弹药并转化为烟云
    for (const auto& [uav_id, uav_strat] : strategy) {
        auto& uav = uavs_.at(uav_id);
        uav.set_flight_strategy(uav_strat.speed, uav_strat.angle);
        for (const auto& g_strat : uav_strat.grenades) {
            auto grenade = uav.deploy_grenade(g_strat.t_deploy, g_strat.t_fuse);
            // 注意：这里不再将烟云与特定导弹绑定
            all_smoke_clouds.push_back(grenade->generate_smoke_cloud());
        }
    }

    if (all_smoke_clouds.empty()) {
        return 1e9;
    }

    // 确定联合模拟的起止时间
    double sim_start_time = all_smoke_clouds[0]->get_start_time();
    double sim_end_time = all_smoke_clouds[0]->get_end_time();
    for (const auto& cloud : all_smoke_clouds) {
        sim_start_time = std::min(sim_start_time, cloud->get_start_time());
        sim_end_time = std::max(sim_end_time, cloud->get_end_time());
    }

    // 为每个导弹计算其被遮蔽的时间（考虑场上所有烟雾云）
    std::unordered_map<std::string, double> total_obscured_time_per_missile;
    for (const auto& missile_id : missile_ids_) {
        total_obscured_time_per_missile[missile_id] = 0.0;
    }

    for (double t = sim_start_time; t < sim_end_time; t += time_step_) {
        // 获取当前时刻场上所有有效的烟云中心
        std::vector<Eigen::Vector3d> active_cloud_centers;
        for (const auto& cloud : all_smoke_clouds) {
            auto center = cloud->get_center(t);
            if (center) {
                active_cloud_centers.push_back(*center);
            }
        }
        
        if (active_cloud_centers.empty()) {
            continue;
        }

        // 对每个导弹独立判断是否被场上所有烟云协同遮蔽
        for (const auto& missile_id : missile_ids_) {
            const auto& missile = missiles_.at(missile_id);
            Eigen::Vector3d missile_pos = missile.get_position(t);
            
            if (Geometry::check_collective_obscuration(missile_pos, active_cloud_centers, target_key_points_)) {
                total_obscured_time_per_missile[missile_id] += time_step_;
            }
        }
    }
    
    // 计算加权总分
    double total_weighted_score = 0.0;
    for (const auto& [missile_id, obs_time] : total_obscured_time_per_missile) {
        total_weighted_score += threat_weights_.at(missile_id) * obs_time;
    }
    
    // 优化器是最小化，所以返回分数的负数
    return -total_weighted_score;
}

std::unordered_map<std::string, double> GlobalOptimizer::calculate_strategy_details(const StrategyMap& strategy) {
    std::unordered_map<std::string, double> total_obscured_time_per_missile;
    for (const auto& id : missile_ids_) {
        total_obscured_time_per_missile[id] = 0.0;
    }

    std::vector<std::unique_ptr<CoreObjects::SmokeCloud>> all_smoke_clouds;
    double min_start_time = 1e9;
    double max_end_time = 0.0;

    for (const auto& [uav_id, uav_strat] : strategy) {
        auto& uav = uavs_.at(uav_id);
        uav.set_flight_strategy(uav_strat.speed, uav_strat.angle);
        for (const auto& g_strat : uav_strat.grenades) {
            auto grenade = uav.deploy_grenade(g_strat.t_deploy, g_strat.t_fuse);
            auto cloud = grenade->generate_smoke_cloud();
            all_smoke_clouds.push_back(std::move(cloud));
            if (all_smoke_clouds.back()->get_start_time() < min_start_time) min_start_time = all_smoke_clouds.back()->get_start_time();
            if (all_smoke_clouds.back()->get_end_time() > max_end_time) max_end_time = all_smoke_clouds.back()->get_end_time();
        }
    }

    if (all_smoke_clouds.empty()) {
        return total_obscured_time_per_missile;
    }

    for (double t = min_start_time; t < max_end_time; t += time_step_) {
        std::vector<Eigen::Vector3d> active_cloud_centers;
        for (const auto& cloud : all_smoke_clouds) {
            auto center = cloud->get_center(t);
            if (center) {
                active_cloud_centers.push_back(*center);
            }
        }

        if (active_cloud_centers.empty()) continue;

        for (const auto& missile_id : missile_ids_) {
            Eigen::Vector3d missile_pos = missiles_.at(missile_id).get_position(t);
            if (Geometry::check_collective_obscuration(missile_pos, active_cloud_centers, target_key_points_)) {
                total_obscured_time_per_missile[missile_id] += time_step_;
            }
        }
    }
    return total_obscured_time_per_missile;
}

} // namespace Optimizer

File: 5.问题五贪心+遗传/optimizer.py
# optimizer.py
from scipy.optimize import differential_evolution
from core_objects import Missile, UAV, TargetCylinder
from config import TRUE_TARGET_SPECS, MISSILES_INITIAL
import numpy as np
from geometry import check_collective_obscuration
class ObscurationOptimizer:
    def __init__(self, missile_id, uav_assignments):
        self.missile = Missile(missile_id)
        # TargetCylinder现在会自动生成完整的关键点集
        self.target = TargetCylinder(TRUE_TARGET_SPECS)
        self.uav_assignments = uav_assignments
        self.time_step = 0.1
        # 预先获取关键点，避免在循环中重复调用方法
        self.target_key_points = self.target.get_key_points()

    def _parse_decision_variables(self, decision_variables):
        # This method MUST be implemented by child classes for specific problems
        raise NotImplementedError

    def objective_function(self, decision_variables):
        strategies = self._parse_decision_variables(decision_variables)
        
        smoke_clouds = []
        try:
            for uav_id, uav_strat in strategies.items():
                uav = UAV(uav_id)
                uav.set_flight_strategy(uav_strat['speed'], uav_strat['angle'])
                for g_strat in uav_strat['grenades']:
                    grenade = uav.deploy_grenade(g_strat['t_deploy'], g_strat['t_fuse'])
                    smoke_clouds.append(grenade.generate_smoke_cloud())
        except (ValueError, KeyError):
            return 0 # Invalid strategy, return worst score

        if not smoke_clouds:
            return 0.0

        sim_start_time = min(c.start_time for c in smoke_clouds)
        sim_end_time = max(c.end_time for c in smoke_clouds)

        obscured_time_points = set()
        for t in np.arange(sim_start_time, sim_end_time, self.time_step):
            # a. 获取当前所有有效的云团中心
            active_cloud_centers = [
                cloud.get_center(t) for cloud in smoke_clouds if cloud.get_center(t) is not None
            ]
            
            if not active_cloud_centers:
                continue

            # b. 调用新的协同判断函数
            missile_pos = self.missile.get_position(t)
            if check_collective_obscuration(
                missile_pos, 
                active_cloud_centers, 
                self.target_key_points # 使用预先缓存的关键点
            ):
                obscured_time_points.add(round(t / self.time_step))
        
        total_time = len(obscured_time_points) * self.time_step
        return -total_time

    def solve(self, bounds, **kwargs):
        result = differential_evolution(
            self.objective_function,
            bounds,
            **kwargs
        )
        optimal_vars = result.x
        max_time = -result.fun
        
        # Re-create the optimal strategy for detailed output
        optimal_strategy = self._parse_decision_variables(optimal_vars)
        return optimal_strategy, max_time


class GlobalOptimizer:
    """
    全局协同优化器：考虑所有导弹同时被场上所有活跃烟雾云遮蔽
    """
    def __init__(self, uav_ids, missile_ids, threat_weights, uav_grenade_counts):
        self.uav_ids = sorted(list(uav_ids))
        self.missile_ids = sorted(list(missile_ids))
        self.threat_weights = threat_weights
        self.uav_grenade_counts = uav_grenade_counts
        
        # 创建所有会用到的对象
        self.missiles = {m_id: Missile(m_id) for m_id in self.missile_ids}
        self.uavs = {u_id: UAV(u_id) for u_id in self.uav_ids}
        self.target = TargetCylinder(TRUE_TARGET_SPECS)
        
        self.time_step = 0.1
        self.target_key_points = self.target.get_key_points()
        self.num_missiles = len(self.missile_ids)

    def _parse_decision_variables(self, dv):
        """
        将一维决策变量向量解析为结构化的全局策略。
        """
        strategy = {}
        dv_index = 0
        
        for uav_id in self.uav_ids:
            num_grenades = self.uav_grenade_counts[uav_id]
            
            # 1. 解析飞行策略
            speed, angle = dv[dv_index : dv_index + 2]
            dv_index += 2
            
            uav_strat = {'speed': speed, 'angle': angle, 'grenades': []}
            
            # 2. 解析每枚弹药的策略
            last_td = 0
            for i in range(num_grenades):
                # 问题五的变量结构
                if i == 0:
                    t_d = dv[dv_index]
                else:
                    # 后续弹药的投放时间是相对于前一枚的增量
                    delta_t = dv[dv_index]
                    t_d = last_td + delta_t
                
                t_f = dv[dv_index + 1]
                
                # 新增：解析该弹药的目标导弹
                # 用一个 0-1 之间的连续变量代表目标选择
                target_selector = dv[dv_index + 2] 
                # 将连续变量映射到离散的导弹索引
                target_missile_index = min(int(target_selector * self.num_missiles), self.num_missiles - 1)
                target_missile_id = self.missile_ids[target_missile_index]

                uav_strat['grenades'].append({
                    't_deploy': t_d, 
                    't_fuse': t_f,
                    'target_missile': target_missile_id
                })
                last_td = t_d
                dv_index += 3 # 每个弹药有3个决策变量: t_deploy/delta_t, t_fuse, target_selector
            
            strategy[uav_id] = uav_strat
            
        return strategy

    def objective_function(self, decision_variables):
        """
        全局目标函数：考虑所有导弹同时被场上所有烟雾云遮蔽
        """
        try:
            strategy = self._parse_decision_variables(decision_variables)
        except (ValueError, IndexError):
            return 1e9 # 返回一个很大的惩罚值，表示这是一个无效的解

        all_smoke_clouds = []
        # 按无人机生成所有弹药并转化为烟云
        for uav_id, uav_strat in strategy.items():
            uav = self.uavs[uav_id]
            uav.set_flight_strategy(uav_strat['speed'], uav_strat['angle'])
            for g_strat in uav_strat['grenades']:
                grenade = uav.deploy_grenade(g_strat['t_deploy'], g_strat['t_fuse'])
                # 注意：这里不再将烟云与特定导弹绑定
                cloud = grenade.generate_smoke_cloud()
                all_smoke_clouds.append(cloud)

        if not all_smoke_clouds:
            return 1e9

        # 确定联合模拟的起止时间
        sim_start_time = min(c.start_time for c in all_smoke_clouds)
        sim_end_time = max(c.end_time for c in all_smoke_clouds)

        # 为每个导弹计算其被遮蔽的时间（考虑场上所有烟雾云）
        total_obscured_time_per_missile = {m_id: 0.0 for m_id in self.missile_ids}

        for t in np.arange(sim_start_time, sim_end_time, self.time_step):
            # 获取当前时刻场上所有有效的烟云中心
            active_cloud_centers = [
                cloud.get_center(t) 
                for cloud in all_smoke_clouds 
                if cloud.get_center(t) is not None
            ]
            
            if not active_cloud_centers:
                continue

            # 对每个导弹独立判断是否被场上所有烟云协同遮蔽
            for missile_id in self.missile_ids:
                missile = self.missiles[missile_id]
                missile_pos = missile.get_position(t)
                
                if check_collective_obscuration(
                    missile_pos, 
                    active_cloud_centers,  # 注意：这里是所有烟云，不分导弹
                    self.target_key_points
                ):
                    total_obscured_time_per_missile[missile_id] += self.time_step
        
        # 计算加权总分
        total_weighted_score = 0
        for missile_id, obs_time in total_obscured_time_per_missile.items():
            total_weighted_score += self.threat_weights[missile_id] * obs_time
            
        # 优化器是最小化，所以返回分数的负数
        return -total_weighted_score

    def solve(self, bounds, **kwargs):
        """
        调用差分进化算法求解。
        """
        result = differential_evolution(
            self.objective_function,
            bounds,
            **kwargs
        )
        
        optimal_vars = result.x
        max_score = -result.fun
        
        # 重新解析最优变量以用于结果展示
        optimal_strategy = self._parse_decision_variables(optimal_vars)
        
        return optimal_strategy, max_score
